### ISSUE TITLE
Fiddling with improvements to counting IDs

### ISSUE THREAD

**Issue by loarie** [inaturalist/inaturalist#530]
We currently count IDs as #IDs made for others. A problem is that this gives equal weight to confirming agrees or coarse rank IDs as it does for IDs that actually add new information. Here I'm fiddling with a now way to count IDs that better reflects novel contributions.

I've broken down a user's IDs into 5 groups:
1) 'novel' IDs as all active IDs where the ID was the earliest ID to equal the observation's community ID
2) 'confirming/coarse' IDs as are all active IDs where the ID was not the earliest ID to equal the observation's community ID
3) 'no_cid' which includes all active IDs on observations that don't yet have a community ID because they are too young or esoteric
4) 'dissents' which are all active IDs that don't match the observation's community ID
5) 'inactive' which are inactive IDs usually because they change their minds or taxon changes

Here's how I'm fetching my novel IDs:
`SELECT count(*)
FROM identifications i
INNER JOIN observations ON (i.observation_id = observations.id)
WHERE i.current = true
AND i.user_id = 477
AND observations.community_taxon_id = i.taxon_id
AND i.created_at = (select MIN(created_at) 
       from identifications i2 where i2.observation_id = i.observation_id);`

This graph shows the breakdown of a few users categories:
![screen shot 2015-04-21 at 8 38 22 am](https://cloud.githubusercontent.com/assets/598026/7256385/9c66d05a-e803-11e4-8229-76034417971f.png)

The top 10 IDers counting by # of 'novel' IDs are: 
finatic (10285) => 19955
greglasley (9706) => 12933
maractwin  (10787) => 10466
kueda (1) => 10059
loarie (477) => 9921
jmaughn (2873) => 7764
dpom (16101_ => 7156
jon_sullivan (1115) => 7061
muir (1000) => 6667
sea-kangaroo (2991) => 6610 

And in terms of number of taxa they've been able to ID by counting # of 'novel' IDs are:
loarie (477) => 3581
kueda (1) => 3310
maractwin (10787) => 2337
finatic (10285) => 2204
greglasley (9706) => 2203
carmelo_lopez (15211) => 2162
sea-kangaroo (2991) => 2030
jon_sullivan (1115) => 1781
jaykeller (27743) => 1747
robberfly (13979) => 1607

The way I've done it here, 'novel' IDs include a user's own IDs as long as they were the first to get to whatever the community ID is. So this isn't a great metric of altruism, but it is a good metric of whether and what people actually know how to ID. 
Here's the same graph from above using only IDs on other people's stuff:
![screen shot 2015-04-21 at 8 38 45 am](https://cloud.githubusercontent.com/assets/598026/7256394/b367897a-e803-11e4-9c3f-3c961450eca9.png)

Here's the top 10 'novel' IDers and species count from novel IDs just for IDs on other people's stuff:

'Novel' IDers:
loarie (477) => 5561
d_kluza (13200) => 3417
kueda (1) => 3060
john8 (11060) => 2556
aztekium (16744) => 2113
maractwin (10787) => 2089
sambiology (22589) => 1388
annemirdl (7504) => 1296
jakob (5643) => 1295
jon_sullivan (1115) => 1259

Species count from novel IDs:
loarie (477) => 2436
kueda (1) => 1532
d_kluza (13200) => 1132
john8 (11060) => 949
aztekium (16744) => 910
maractwin (10787) => 811
jon_sullivan (1115) => 749
sambiology (22589) => 677
meurkc (5302) => 656
rcurtis (4860) => 649

**Comment by alexshepard** [comment]
A couple of things:

 * I'm curious how we'll quantify this number to users? Does this basically become a "IDer Score?"
 * Does this account for IDs on my observation vs IDs of others observations? I.e. is finatic a prolific IDer of other people's observations? Or is his "high score" here just a corollary of the fact that he observes so much and he IDs what he observes? If we're trying to quantify the impact an IDer has on the community at large, it might be meaningful to make this distinction.
 * When we did the push to a million observations, there was a concerted effort to get observations to research grade. This metric system doesn't reward that. I'm curious if there's a way to reward the adding of an ID that promotes an observation to research grade. I'm especially curious if this is do-able if an observation can be promoted to research grade at a coarse ID and then is refined later.

As we go down the gamification pathway here, it's also worth keeping in mind how metrics like this have shaped communities. Think karma at slashdot and stack overflow's reputation. http://meta.stackexchange.com/questions/17204/six-simple-tips-to-get-reputation-fast-on-any-stack-exchange-site

**Comment by loarie** [comment]
Alex - I agree that we don't want to accidentally disincentivize altruism (e.g. IDing other people's obs), confirming IDs, or coarse/quick IDs. So I'm not advocating this as a quick swap out of the current system. But I think its interesting because it more accurately captures the species people have learned. Which might be useful for showing them obs to confirm in something like a goals tab or update feed to better get things to Research Grade. It could also be a kernel of a reputation system that would allow people to get photoless obs to Research Grade if they are of species they've shown they have learned. For example, based on this way of counting, these are taxa you've learned:  
Arthropods (Arthropoda) 
Spiders (Araneae)
Insects (Insecta)
Ladybird Beetles (Coccinellidae)
Earwigs (Dermaptera)
Butterflies and Moths (Lepidoptera)
Monarch Butterfly (Danaus plexippus)
Red Admiral (Vanessa atalanta)
Cabbage White (Pieris rapae)
Plants (Plantae)
Orchids (Orchidaceae)
Birds (Aves)
Ospreys (Pandionidae)
Crows, Jays, and Magpies (Corvidae)
Human (Homo sapiens)
California Sea Lion (Zalophus californianus)
Eastern Gray Squirrel (Sciurus carolinensis)
Douglas Squirrel (Tamiasciurus douglasii)
Spotted Salamander (Ambystoma maculatum)
Banana Slugs (Ariolimax)
Fungi including Lichens (Fungi)
Toothed Jelly Fungus (Pseudohydnum gelatinosum)

**Comment by naturalistcharlie** [comment]
I like this. Don't have much to add now because I'm in Costa Rica
identifying sloths and stuff :). However there is one sort of id worthy of
mention. Say I am in Costa Rica and see a bird and I have no clue what it
is. Scott says it is a resplendent quetzal. I have no reason to disbelieve
him so I agree. So then it's research grade even if i don't know a
resplendent quetzal from a pigeon (I'm not quite that bad at birds but for
the sake of example). Not sure if there is a way to address this but I'd
rather agree with such an id than wait for someone else to agree so...

**Comment by tnywlls** [comment]
Interesting.  I've often thought that a user-expertise weighting based on how often they were the first to 'correctly' id an observation would be useful.  At present my ids, those of a 5 year old and those of a recognised expert are all given equal weight which seems odd to say the least.  Sometimes when someone has initially incorrectly id'd something it takes 3 correct ids to fix it.  Such a weighting would allow recognised 'experts' to make research grade observations almost automatically (as they would in most other systems), and allow others to prove their unrecognised expertise.  Perhaps such a weighting shouldn't be retrospective, ie you don't go and rescore old identifications by someone who has now proven their expertise as they may have only just reached that level and earlier ids might not be so reliable. And you probably don't want to continuosly recalculate the score for every observation anyway.

**Comment by kueda** [comment]
I like this as a way of incentivizing what you've described as novel identifications. I might propose this renaming, though:

1. <s>'novel'</s> **lead** IDs as all active IDs where the ID was the earliest ID to equal the observation's community ID **[because "novel" suggests "first" or "early" to me]**
1. '**confirming**<s>/coarse</s>' IDs as are all active IDs where the ID was not the earliest ID to equal the observation's community ID **[this means all active IDs where the ID mathces the CID AND is not the earliest, right?]**
1. <s>'no_cid'</s> **novel** which includes all active IDs on observations that don't yet have a community ID because they are too young or esoteric **["putative" or "preliminary" might be more accurate but also longer and more technical]**
1. <s>'dissents'</s> **disagreeing** which are all active IDs that don't match the observation's community ID **[longer, but plainer]**
1. 'inactive' which are inactive IDs usually because they change their minds or taxon changes

Technically, this would be a bit of a pain if we stored these statuses with the identifications, but I don't see a better way to do it. Every time a new identification was added we'd need to re-determine the type for each one, though frankly we already do something similar with active/inactive IDs. We'd also have to retroactively classify all observations.

Implementation-wise, I'm think just adding a some kind of `idtype` string column to `identifications` that would be `novel` by default, but could also be `lead`, `confirming`, or `inactive`, and displaying these types on the list of identifications on obs/show, with the future intent of showing some of the kinds of stats @loarie has put together here. Does that sounds reasonable, @pleary, or do you think these things can be determined dynamically?

**Comment by loarie** [comment]
>'novel' lead IDs as all active IDs where the ID was the earliest ID to equal the observation's community ID [because "novel" suggests "first" or "early" to me]

sounds good to me
>'confirming/coarse' IDs as are all active IDs where the ID was not the earliest ID to equal the observation's community ID [this means all active IDs where the ID mathces the CID AND is not the earliest, right?]

coarse was meant to capture ancestors - for example, if the ID chain goes (1) Animals, (2) Birds, (3) Birds the CID is Birds. Animals would fall in this bin because while it doesn't match the CID, its still an ancestor of the CID

>'no_cid' novel which includes all active IDs on observations that don't yet have a community ID because they are too young or esoteric ["putative" or "preliminary" might be more accurate but also longer and more technical]

sounds good to me
>'dissents' disagreeing which are all active IDs that don't match the observation's community ID [longer, but plainer]

sounds good to me

My main question is we were to change what kind of ID's get 'counted' should we 'count' ancestors (ie Coarse above) - e.g. if the ID chain goes (1) Animals (SL), (2) Birds (KU), (3) Birds (SL) should the only 'point' awarded be your (3) as the 'lead' ID or since my (1) was the 'lead' briefly before I added (3) should I also get a point, or should it be revoked upon (3)?

**Comment by kueda** [comment]
I feel like a **coarse** ID is worth more than a **confirming** ID, unless that **confirming** ID changed the quality grade of the observation, which would sort of add another category to the scheme. To me, the only IDs that should not "count" are the redundant confirmations. Categorizing them as "redundant" seems unnecessarily mean, though.

**Comment by naturalistcharlie** [comment]
I do think those 'redundant' IDs do have some value (at least the third,
fourth one, maybe not after that).... for several reasons for instance
cases where someone IDs something for me and I agree just because it looks
right, but am not an expert on that species... would be good to get a third
ID.  However, I think sometimes people agree just to mark that they saw
something, even if many others have seen it... so the feature to mark
things as reviewed should help.

On Thu, Jun 25, 2015 at 4:22 PM, Ken-ichi <notifications@github.com> wrote:

> I feel like a *coarse* ID is worth more than a *confirming* ID, unless
> that *confirming* ID changed the quality grade of the observation, which
> would sort of add another category to the scheme. To me, the only IDs that
> should not "count" are the redundant confirmations. Categorizing them as
> "redundant" seems unnecessarily mean, though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/inaturalist/inaturalist/issues/530#issuecomment-115384922>
> .
>



-- 
============================
Charlie Hohn
Montpelier, Vermont
coyot.es/slowwatermovement

**Comment by loarie** [comment]
I just reread how I was calculating these bins and realized that my comments above on 'confirming/coarse'  don't make sense. As I had it, the Confirming category *does not* include IDs of ancestors (e.g. coarse) of the community ID, those would be in the Disagree category. Was confusing myself... Here's the algorithms I used to calculate the categories (now using Ken-ichi's names):

* Lead ID (first id matching Community ID)
```
SELECT count(*) FROM identifications i
INNER JOIN observations ON (i.observation_id = observations.id)
WHERE i.current = true
AND i.user_id = 477
AND observations.community_taxon_id = i.taxon_id
AND i.created_at = (
    SELECT MIN(created_at)
    FROM identifications i2
    WHERE i2.observation_id = i.observation_id
);
```
* Confirming ID  (id matching Community ID but not first)

*this includes both IDs that alter the community ID and those that don't, should they be separated?*
```
SELECT count(*) FROM identifications i
INNER JOIN observations ON (i.observation_id = observations.id)
WHERE i.current = true
AND i.user_id = 477
AND observations.community_taxon_id = i.taxon_id
AND i.created_at != (
    SELECT MIN(created_at) 
    FROM identifications i2
    WHERE i2.observation_id = i.observation_id
);
```
* Novel ID  (id on an obs with no Community ID)
```
SELECT count(*)
FROM identifications
INNER JOIN
observations ON (identifications.observation_id = observations.id)
WHERE identifications.user_id = 477
AND identifications.current = true
AND observations.community_taxon_id IS NULL;
```
* Disagree (ids not matching the Community_id)

*this includes IDs of ancestors and descendants of the community IDs as well as cousins, should they be separated?*

```
SELECT count(*) 
FROM identifications
INNER JOIN
observations ON (identifications.observation_id = observations.id)
WHERE identifications.user_id = 477 
AND identifications.current = true
AND observations.community_taxon_id != identifications.taxon_id;
```
* Inactive (ids of inactive taxa)
```
SELECT count(*)
FROM identifications
INNER JOIN
observations ON (identifications.observation_id = observations.id)
WHERE identifications.user_id = 477
AND identifications.current = false;
```

I think we all agree that we should generate notifications for all IDs except confirming-IDs-that-don't-alter-the-community-ID ('redundant confirming IDs').

Regarding how to 'count' these, is it to radical to *only* count lead IDs? For example, I have 10,757 lead IDs of 42,807 total IDs. IMO the advantage of counting lead IDs is we'd only reward IDs that move the community ID forward and turn out to be correct, e.g. there'd be no free riding with confirming IDs/bad IDs to increase your count. The disadvantage is there'd be a bit of a lag in an IDers count increasing until the community ID kicked in, would folks complain? Likewise would this disincentivize 'non-redundant confirming IDs'?

Ken-ichi above mentioned counting all IDs except 'redundant confirming IDs', advantage here is most IDs are counted/rewarded/incentivized, disadvantage is there's quite a bit of free-riding with confirming IDs and Disagrees that are actually wrong so we wouldn't be rewarding based on IDs that turn out to be correct. So ID count wouldn't be as good a proxy for reputation.

Thoughts?

**Comment by loarie** [comment]
@kueda what are your feelings on 'Regarding how to 'count' these, is it to radical to only count lead IDs?' above? Presuming the primary use case for this would be IDer leaderboards 
https://github.com/inaturalist/inaturalist/issues/640

I'm a fan of only counting lead IDs because its better representation of people that are actually doing the heavy lifting on IDing (vs agreeing), but I wouldn't want to de-incentivize agreeing.

**Comment by kueda** [comment]
Well, you've already highlighted two big problems with only counting lead IDs:

1. Lag time: we won't know it's a lead ID until someone confirms it, which means if you're tracking your stats closely, you won't be able to impact them by simply adding correct IDs
1. De-incentivize agreeing: we do want agreements, and only counting leads completely devalues them.

I think those are both pretty big downsides, especially the first if we're trying to gamify things a bit more (shudder), since it will feel a bit pointless if you have to wait a few weeks to see your time spent identifying affect your ID count. I guess that's kind of an argument for a score instead of a count, e.g. lead ID counts for 2 points, first confirming IDs count for 1, everything else doesn't contribute to your score.

**Comment by loarie** [comment]
Just wanted to get these notes down so I don't forget them. I have no idea if there's a way to make this performative, but IMO, this is the best way to count IDs:
```
#Select all IDs for a user
#that are of the community_id (or any of the community_ids ancestors)
#and are the first time within an observation
#that that taxon (or any of its descendants) were proposed
sql = "SELECT i.id, i.taxon_id, t3.ancestry FROM identifications i
INNER JOIN observations ON (i.observation_id = observations.id)
INNER JOIN taxa t1 ON (observations.community_taxon_id = t1.id)
INNER JOIN taxa t3 ON (i.taxon_id = t3.id)
WHERE i.user_id = 477
AND ( 
  observations.community_taxon_id = i.taxon_id
  OR t1.ancestry LIKE '%/' || i.taxon_id || '/%'
  OR t1.ancestry LIKE '%/' || i.taxon_id
)
AND i.created_at = (
  SELECT MIN(i2.created_at)
  FROM identifications i2
  INNER JOIN taxa t2 ON (i2.taxon_id = t2.id)
  WHERE i2.observation_id = i.observation_id
  AND (
    i2.taxon_id = i.taxon_id
    OR t2.ancestry LIKE '%/' || i.taxon_id || '/%'
    OR t2.ancestry LIKE '%/' || i.taxon_id
  )
)"
results = ActiveRecord::Base.connection.execute(sql)
results.count #18294
taxon_ids = results.map{|id| id["taxon_id"]}.uniq
taxon_ids.count #4997
ancestors = results.map{|id| id["ancestry"]}.join("/").split("/").uniq
leaves = taxon_ids.reject{|taxon_id| ancestors.include? taxon_id}
leaves.count #3991
```
These are all IDs that based on the community ID moved the ball forward. (Lets call them 'improving IDs' for now). For example, here those IDs would be Reptiles and Trachemys:
<img width="653" alt="screen shot 2015-11-06 at 4 50 54 pm" src="https://cloud.githubusercontent.com/assets/598026/11012613/83223e3c-84ab-11e5-99d5-adc2ed4f44d5.png">
Map turtle is not supported by the community ID, Testudines is an ancestor of a previous suggestion (Trachemys) so does not count, Trachemys scripta is not (yet) supported by the community ID

Whats cool about this is that the taxa that the IDs represent are taxa that the user knows how to identify. So this observation tells us that thehapplesshiker knows how to ID Reptiles, so we can show her other Reptile obs needing confirmation. Likewise pinkpimpernel knows how to ID Trachemys so we can show her other Trachemys obs (and presumably ancestors of Trachemys) needing confirmation. 

It also has a nice parallel with obs: You've made X obs representing Y species (ahem leaves). We can say you've made Z identifications representing W species (ahem leaves) - with the implication being that you've learned how to ID those species  which opens up all sorts of interesting avenues (just like implication of your lifelist is that you've seen those species). Likewise, there's analogous caveats. For observations, we can say that for 'the rules of the game' (ie the stats that we compare across the community) we'll only count research,needs_id obs. And for IDs, we can say we only count these 'improving IDs'. Ideally, individual users and projects could opt to rather count all obs/IDs, but for comparisons across the community (e.g. competition) would be nice to have only obs/IDs that are meaningful like these counted.

A drawback of this approach we've discussed is that there's a delay since we're judging the merits of an ID based on the community ID. I've thought about this alot, and still think this is OK across the community (ideally if individuals could opt out if they wanted to focus on their raw count).

A second potential drawback I haven't had a chance to think about yet is that if these IDs counted this way can be used to calculate the taxa people know how to ID, its a little weird that one correct ID will get this taxon counted, but many incorrect IDs won't cancel this out. (e.g. If I made an 'improving ID' of an American robin once, it would go on my list of taxa I can ID, but if I misID'd 100 starlings as American robins, this wouldn't be considered). Have to think about this more to make sense of it, but just wanted to make a note of this concern.

**Comment by loarie** [comment]
Ok - I think I've thought all of this through. First, I propose a quick change and remove the ability to delete IDs - instead, you’ll be able to make them not current - but you won’t be able to delete them.

With that, there are 5 kinds of identifications:
### Improving IDs
The first are ‘improving IDs’ which are IDs that move the ball forward
```
#Select all IDs for a user
#that are of the community_id (or any of the community_ids ancestors)
#and are the first time within an observation
#that that taxon (or any of its descendants) were proposed
sql = "SELECT i.id, i.taxon_id, t3.ancestry FROM identifications i
INNER JOIN observations ON (i.observation_id = observations.id)
INNER JOIN taxa t1 ON (observations.community_taxon_id = t1.id)
INNER JOIN taxa t3 ON (i.taxon_id = t3.id)
WHERE i.user_id = 477
AND ( 
  observations.community_taxon_id = i.taxon_id
  OR t1.ancestry LIKE '%/' || i.taxon_id || '/%'
  OR t1.ancestry LIKE '%/' || i.taxon_id
)
AND i.created_at = (
  SELECT MIN(i2.created_at)
  FROM identifications i2
  INNER JOIN taxa t2 ON (i2.taxon_id = t2.id)
  WHERE i2.observation_id = i.observation_id
  AND (
    i2.taxon_id = i.taxon_id
    OR t2.ancestry LIKE '%/' || i.taxon_id || '/%'
    OR t2.ancestry LIKE '%/' || i.taxon_id
  )
)"
results = ActiveRecord::Base.connection.execute(sql)
results.count #18297 
ancestors = results.map{|id| id["ancestry"]}.join("/").split("/").uniq
taxa = results.map{|id| id["taxon_id"]}.group_by{|i| i}.map{|k,v| {:taxon_id => k.to_i, :count => v.count} }
leaves = taxa.reject{|entry| ancestors.include? entry[:taxon_id].to_s}.sort_by{|i| i[:count]}.reverse
leaves.count #3,987
leaves[0..5]
```
These tell us what species you are able to ID. For example,I’ve made improving IDs (i.e. the community Ids vindicated that I correctly ID on my own) Western Fence Lizard 82 times, and Common Yarrow 74 times:
```
[{:taxon_id=>36204, :count=>82}, {:taxon_id=>52821, :count=>74}, {:taxon_id=>120720, :count=>69}, {:taxon_id=>48225, :count=>60}, {:taxon_id=>47219, :count=>57}...
```
This list of species should be available to the user and they should be able to humbly click ‘I actually don’t know how to ID this’ and be able to remove things from this list by making the IDs behind them no longer current. Using the above query to look at my own list of taxa I ‘can ID’, over the years I’ve often added an ID of the person’s placeholder taxon when I actually don’t have a clue how to ID the taxon. This is meant to be a friendly tip to the new user until they get the hang of how IDs work, and I usually tell myself that I’ll remove my ID after they ‘agree’. But clearly I often don’t. Here’s an example http://www.inaturalist.org/observations/85097. I wouldn’t know a two-spotted bumblebee if it stung me in the ass. The responsible thing for me to do would be to review this list from time to time and remove things I really can’t ID, and this feature would help me do it.
<img width="993" alt="screen shot 2015-11-06 at 10 36 33 pm" src="https://cloud.githubusercontent.com/assets/598026/11014053/262db550-84db-11e5-8626-1661d0d6bda9.png">

### Maverick IDs
The second group are ‘maverick IDs’ these are IDs that are not ancestors of or descendants of the community ID.
```
#ids that are NOT ancestors of or descendants of the community_id
sql = "SELECT i.id, i.taxon_id, t3.ancestry FROM identifications i
INNER JOIN observations ON (i.observation_id = observations.id)
INNER JOIN taxa t1 ON (observations.community_taxon_id = t1.id)
INNER JOIN taxa t3 ON (i.taxon_id = t3.id)
WHERE i.user_id = 477
AND i.current = true
AND ( 
  observations.community_taxon_id != i.taxon_id
  AND t1.ancestry NOT LIKE '%/' || i.taxon_id || '/%'
  AND t1.ancestry NOT LIKE '%/' || i.taxon_id
  AND t3.ancestry NOT LIKE '%/' || observations.community_taxon_id || '/%'
  AND t3.ancestry NOT LIKE '%/' || observations.community_taxon_id
)"
results = ActiveRecord::Base.connection.execute(sql)
results.count #18294
```
Because these all have a community ID which is different than what I’m doing - in these cases I’m that one kid who’s doing his own thing. Most often these are IDs I can’t believe I actually made, am super embarrassed about, and would love a tool to fix as quickly as possible before someone calls me out in public, like this one: http://www.inaturalist.org/observations/136452
<img width="974" alt="screen shot 2015-11-06 at 10 23 53 pm" src="https://cloud.githubusercontent.com/assets/598026/11014063/55a0d056-84db-11e5-93d2-c875005776dd.png">

But some are ones where I really am a Maverick like this one, and I should be motivated to take up arms in these cases to fight the power
http://www.inaturalist.org/observations/275338

<img width="960" alt="screen shot 2015-11-06 at 10 22 11 pm" src="https://cloud.githubusercontent.com/assets/598026/11014065/84202eb8-84db-11e5-970d-218a27796fa6.png">

In either case a place on my dashboard reminding me of these IDs would be very useful
### Leading IDs
Next, I’m going to take back the term ‘leading IDs’ for IDs that are descendants of the community ID, or where there is no community ID.
```
#ids that are descendants of the community_id or no communityID
sql = "SELECT i.id, i.taxon_id, t3.ancestry FROM identifications i
INNER JOIN observations ON (i.observation_id = observations.id)
INNER JOIN taxa t3 ON (i.taxon_id = t3.id)
WHERE i.user_id = 477
AND i.current = true
AND observations.quality_grade IN ('research', 'needs_id')
AND (
  observations.community_taxon_id IS NULL
  OR t3.ancestry LIKE '%/' || observations.community_taxon_id || '/%'
  OR t3.ancestry LIKE '%/' || observations.community_taxon_id
)"
results = ActiveRecord::Base.connection.execute(sql)
results.count #18294
results.each do |r|
  puts Identification.find_by_id(r["id"]).observation_id
end
```
These observations need attention by someone other than me. Take my Baldhip rose obs: http://www.inaturalist.org/observations/682892.
<img width="977" alt="screen shot 2015-11-06 at 11 10 42 pm" src="https://cloud.githubusercontent.com/assets/598026/11014067/be29addc-84db-11e5-95d4-12be7c0c0574.png">
I’ve been waiting a year for someone to agree with me on this damn thing. But who?? This is where the ‘Improving IDs’ from above come in:
```
taxon_id = 53441
sql = "SELECT i.id, i.user_id, i.taxon_id, t3.ancestry FROM identifications i
INNER JOIN observations ON (i.observation_id = observations.id)
INNER JOIN taxa t1 ON (observations.community_taxon_id = t1.id)
INNER JOIN taxa t3 ON (i.taxon_id = t3.id)
WHERE i.taxon_id = #{taxon_id}
AND ( 
  observations.community_taxon_id = i.taxon_id
  OR t1.ancestry LIKE '%/' || i.taxon_id || '/%'
  OR t1.ancestry LIKE '%/' || i.taxon_id
)
AND i.created_at = (
  SELECT MIN(i2.created_at)
  FROM identifications i2
  INNER JOIN taxa t2 ON (i2.taxon_id = t2.id)
  WHERE i2.observation_id = i.observation_id
  AND (
    i2.taxon_id = i.taxon_id
    OR t2.ancestry LIKE '%/' || i.taxon_id || '/%'
    OR t2.ancestry LIKE '%/' || i.taxon_id
  )
)"
results = ActiveRecord::Base.connection.execute(sql)
results.map{|id| id["user_id"]}.group_by{|i| i}.map{|k,v| {:user_id => k.to_i, :count => v.count} }.sort_by{|i| i[:count]}.reverse
```
Turns out there are 20 people on the site who ‘know how to ID’ (e.g. made ‘improving IDs’ of) Baldhip rose, and this is how many times they’ve done it:
```
[{:user_id=>28920, :count=>3}, {:user_id=>38530, :count=>2}, {:user_id=>41234, :count=>2}, {:user_id=>1, :count=>2}, {:user_id=>13377, :count=>2}, {:user_id=>933, :count=>1}, {:user_id=>628, :count=>1}, {:user_id=>82296, :count=>1}, {:user_id=>20912, :count=>1}, {:user_id=>2873, :count=>1}, {:user_id=>104298, :count=>1}, {:user_id=>462, :count=>1}, {:user_id=>45740, :count=>1}, {:user_id=>20035, :count=>1}, {:user_id=>6637, :count=>1}, {:user_id=>101988, :count=>1}, {:user_id=>87323, :count=>1}, {:user_id=>477, :count=>1}, {:user_id=>5327, :count=>1}, {:user_id=>16101, :count=>1}]
```
Imagine that on the observation it had a ‘try asking these folks’ section with links to their profiles

Likewise, on their dashboards, there could be a section that would encourage them to take a look at these observations (alongside a humble ‘I actually don’t know how to ID this’ button).

I also suggest we put an ‘are you sure’ warning anytime someone wants to add an ID that is an ancestor of other IDs. e.g. the obs has an ID of Acer negundo and I add an ID of Acer. This is, admittedly confusingly, meant to imply that I think it is NOT Acer negundo. But this is most often done by people thinking ‘I know its Acer and having nothing to say about the species’. An ‘are you sure’ warning explaining this would surely cut down on this. And shouldn’t be too inconvenient since it’s rare, deliberate, and cathartic when people actually do mean ‘this is NOT Acer negundo’.

And if we are going to start rallying people on their dashboards to ID certain taxa, this would insure against this scenario: Frank can ID Acer so we show him an obs ID’d to Acer on his dashboard and say ‘can you confirm this observation’. He clicks on it, but by that time its already ID’d to Acer negundo. Frank is just following orders, he knows its Acer but doesn’t know anything about Acer negundo so makes an ID of Acer. Up comes the alert “This implies that you think that this observation is NOT of Acer negundo, but rather something else in Acer. Is this the correct? Yes, Cancel”
### Supporting IDs
These are IDs that like ‘improving IDs’ match the community ID (or an ancestor of the community ID), but they aren’t the first. We like them mainly because they make things research grade. And since they no longer necessarily generate updates, they do no harm.
### Not current IDs
These are IDs that have been crossed out because they’ve been changed or (see proposal above) removed. They do no harm but are a useful log of how the identification process unfolded.

**Comment by carrieseltzer** [comment]
I've been meaning to say ever since I read this last week that I love love love everything about this. Just one small detail clarification question--instead of deleting ids, you'd be able to click something like "not current" instead? I'm thinking of a case where I can only id to some coarse level but it progressed to a species id that I'm not comfortable agreeing with so in this new system I'd click "not current" to keep my old id from dragging down the community id (rather than deleting it). Is that right? 

I would love to see #641 (set obs taxon_id when nil with any user's id) tackled at the same time as this.

**Comment by joellebel** [comment]
This all sounds good to me. My only concern is showing the user all of these different types of IDs in their user profile, but we've talked Scott and I think you agree. All of the stats don't necessarily have to be presented simply as numbers, but rather we can use them to surface relevant material to people. 

Since I'm designing the Identification tool, now would be a good time to weave some of these stats in the that ui. I will explore a couple different directions.

**Comment by loarie** [comment]
Slight tweaks to the above definitions. Categories are 'leading', 'improving', 'maverick', and 'supporting' (these are independent of whether an ID is current or not)

The logic is:
```
if o.community_taxon.blank? || descendant_of_community_taxon
  categories[:leading] << ident
elsif ( ancestor_of_community_taxon || matches_community_taxon ) && progressive
  categories[:improving] << ident
elsif !ancestor_of_community_taxon && !descendant_of_community_taxon && !matches_community_taxon
  categories[:maverick] << ident
else
  categories[:supporting] << ident
end
```
where 'progressive' means was an ID the first to move the Community ID (CID) in the direction of the current CID. For example if the ID chain was (1) Insects, (2) Insects, (3) Beetles, (4) Butterflies, (5) Beetles, (6) Beetles, (7) Beetles, then IDs 1 & 3 would be progressive
 
In English this means:

Leading: Any ID that is a descendant of the CID (or is ID on an observation with no CID)
Improving: Any ID that matches the CID or its ancestors that was progressive
Maverick: Any ID that does not match the CID or its ancestors or descendants
Supporting: Any ID not matching the above categories. Alternatively, can be thought of as 'non-progressive' Improving IDs, ie IDs that match the CID or its ancestors but were not the first, but supported an improving ID.

**Comment by carrieseltzer** [comment]
I know this thread has largely been about how we credit IDs to users, but I have also been thinking about how IDs for observations progress in general and how it might be improved. Now that it takes just one user who isn't the observer to set an ID (rather than 2 as it used to be), I've definitely changed my behavior.

Example:
I add a coarse ID "beetles" to one with no ID. Then borisb finds it and adds a genus ID. The CID is still beetles. I don't have time/interest/expertise to confirm the finer ID but I don't want my coarse ID to stand in the way of the observation being found and confirmed by someone else who knows that beetle genus. So, I delete my coarse ID so that the ID shows as the genus instead of beetles. I probably do this 3-5 times a day lately, sometimes with my own observations.

I recall the discussion of an option to inactivate rather than delete an ID, which I would absolutely use and prefer to deletion. But another option may be to have the searchable ID jump to the lead ID in cases where it's a descendant of an earlier ID. Then the onus wouldn't be on me to inactivate my earlier coarse ID (in the example above).

A similar idea may have been proposed in the Google Group a couple months back but I'm composing this on a plane without Internet. Also happy to bring this to the Google groups of you think that would be helpful.

**Comment by carrieseltzer** [comment]
Funny, I didn't read yesterday's discussion on this exact topic in the Google Group until just now (because I get the daily digest). The idea must have been in the air! l'll add my comments there and no need to hijack this thread that is about "scoring" IDs.

**Comment by loarie** [comment]
@kueda , checking in on the idcats stuff. For me after playing with it quite a bit this last month this infrastructure seems useful, what are your thoughts?

If you agree its useful, we should think about how to make use of it in the site. Maybe a good place to start would be on the observations page:
1) People who can help (essentially as you've implemented it as a test)
2) Attaching an 'Earned reputation' metric to IDs (e.g. # of improving IDs by that user for that taxon) as a way to indicate the 'trustworthyness' of IDs (e.g. an ID by someone who's made >5 improving IDs of that taxon would be decorated differently than and ID by someone who's made 0)

Thoughts on these? Or alternative strategies for moving this forward?

**Comment by kueda** [comment]
I think I'd prefer to start by just showing the categories on the obs page, but I like the direction of 2. Maybe instead of popover we have now when you hover over IDs we could show some details about how good that user is at identifying that taxon. I think that would be pretty quick and easy, but we would need to get this stuff in ES. Want to write up a functional spec? This could also wait on a redesign of obs detail.

**Comment by joellebel** [comment]
I personally would like to do include this with the ob detail which it sounds like will come soon after species page.

**Comment by kueda** [comment]
If you want to start doing things with this data, please make new issues.

**Comment by johnnybirder** [comment]
Long discussion! Pretty cool. My one comment on this. There is a current subspecies IDing wave taking over Africa (I'm one of them so I'm not complaining). However, subspecies are very often IDd by distribution; the hard work really was in the species ID. Many of the subspecies IDers aren't actually very good in higher-level taxon IDs, but they still get the community to "corral" around their IDs (as described above). So, ideally, I would like to see the above ratings to reflect the actual work people put in, thus, ignoring IDs say above family (e.g. those just classifying a "something" into a basket) and below species (e.g. subspecies), except in cases where the ID jumped straight to the subspecies ID (i.e. skipping species/family ect.)
